### PR TITLE
Plugin manager usability improvements

### PR DESCRIPTION
**Pull request by eht16**
This PR makes two smaller improvements to the plugin manager dialog:

a) enable interactive search in the plugin list to quickly search the list (which is nowadays quite long, yay)
b) add a popup menu to each plugin to quickly access the configure, keybindings and help functions

### CODE REVIEW

**Comment by b4n** [comment]
LGTM, apart maybe the small details I commented on

**Comment by b4n** [comment]
Nice features, too :)

**Comment by kugel-** [comment]
My opinion is that an always-visible search that is always visible (or which you invoke explicitly with Ctrl+F) should perform a richer search, i.e. substring search or search in the description, whereas an implicit type-ahead-find-like search should only perform prefix-match in the major column (just like in the keybinding dialog or in the sidebar).

And I would prefer the latter type-ahead search for the plugin manager if I had to choose, but they should be able to co-exist.

**Comment by techee** [comment]
@kugel- I agree totally - I think the prefix-only search is something one expects from the quick search feature Enrico implemented and I think the patch is fine like this.

And yes, the always-visible search would be best to search in substrings and filter the list to contain just the matching rows. But as you said, both of the searches can exist in parallel because they serve a slightly different purpose.

**Comment by eht16** [comment]
I implemented a new variant, mostly as suggested before:
a always visible search box with some kind of rich search method: any substring will match and the search is performed against plugin name *and* description.

Since even this always visible search box is based and therefore tight to the interactive search feature of the treeview, I don't know whether it would be possible to enable the "original" interactive search seperately. Maybe we could add a keypress event handler on the treeview to intercept keypresses and redirect them to the search box. Will try later after I pushed my recent changes.

**Comment by eht16** [comment]
Type-a-head search enabling by redirecting the key press from the treeview to the search entry actually worked. Implemented in f520741, hopefully properly :).

I connected the key-press-event signal handler with g_signal_connect_after(), in order to let the GtkTreeView handler process the event first in case it is some special key combination or something which is not intended to trigger an interactive search.

By the way, the interactive search as well as the new search entry have a nice feature: you can navigate through matches with cursor UP/DOWN and also Ctrl-g/G and trigger the interactive search with Ctrl-f. Don't know if this is widely known, I just learned it by reading the GTK source code :).

For testing, you could search for terms like "port", "debug" or "plugin", all match a couple of plugins in plugin names as well as descriptions.

**Comment by b4n** [comment]
From a UI perspective IMO @techee suggestion is better for an always-visible search entry -- filtering.  But as it's more work, we might want to just have what you wrote here, and see later if we want more.

Also, while I agree that putting the search entry next to the buttons avoids wasting space, I'm not sure it's a great place because, well, one would expect buttons there.  I think a better UI would be to have all four buttons next to each other with the search entry above the list:

```
 Select the plugins you want to enable
 
 [.Search......................................]
 
 /---------------------------------------------\
 |                                             ^
 | PLUGIN ONE                                  #
 | Some nice description                       #
 |                                             #
 | PLUGIN TWO                                  #
 | Another long description full of interestin #
 |                                             #
 | PLUGIN THREE                                #
 | This plugin is very important               #
 |                                             #
 | PLUGIN FOUR                                 #
 | Yet another invaluable plugin you must inst #
 |                                             #
 | PLUGIN FIVE                                 #
 | Ok, the other plugins are useful but this o v
 \_____________________________________________/
 
 [ Help ] [ Preferences ] [ Keybindings ] [ OK ]
```

capture du 2015-02-14 00 10 32
*(real world example, but without the search field for now because I did this from an old test I had)*

**Comment by b4n** [comment]
> By the way, the interactive search as well as the new search entry have a nice feature: you can navigate through matches with cursor UP/DOWN and also Ctrl-g/G and trigger the interactive search with Ctrl-f. Don't know if this is widely known, I just learned it by reading the GTK source code :).

I knew about `^F`, `↑` and `↓` but not `^g` and `^G` :)
Though, as nice as `↑` and `↓` are, they are not very discoverable, which is partly why I think a filter is more user friendly (and it also removes the crowing allowing to focus only on the matches).

**Comment by eht16** [comment]
I don't mind dropping the implementation of the key-press event redirection as I agree it is somewhat fragile and might have side effects we don't know yet.
I played with it mostly for my curiosity and because others suggested to have both ways of search in parallel. 

I do like the idea of putting the search entry above the list (I thought about it first but then used the empty space next to the buttons) and join the plugins with the OK button as on your screenshot.
I will implement this soonish.

**Comment by b4n** [comment]
> I do like the idea of putting the search entry above the list (I thought about it first but then used the empty space next to the buttons) and join the plugins with the OK button as on your screenshot.
I will implement this soonish.

I you like I can cleanup my "joined buttons" changes and make a PR?

**Comment by eht16** [comment]
Too late :).
I just pushed a new version with a filter entry, I actually like the idea much, and incorporated the joined buttons idea.

This is how it looks like now: 
plugin_manager_dialog

File: src/plugins.c [hunk 1]
```diff
@@ -1516,36 +1639,44 @@ static void pm_show_dialog(GtkMenuItem *menuitem, gpointer user_data)
...
 	gtk_box_pack_start(GTK_BOX(vbox2), swin, TRUE, TRUE, 0);
-	gtk_box_pack_start(GTK_BOX(vbox2), hbox, FALSE, TRUE, 0);
 
 	g_signal_connect(pm_widgets.dialog, "response", G_CALLBACK(pm_dialog_response), NULL);
 
 	gtk_box_pack_start(GTK_BOX(vbox), vbox2, TRUE, TRUE, 0);
 	gtk_widget_show_all(pm_widgets.dialog);
+	gtk_widget_show_all(pm_widgets.popup_menu);
+
+	gtk_widget_grab_focus(pm_widgets.tree);
```
Comment (eht16) [inline src/plugins.c]: Just for the records: grabbing the focus was necessary to not leave the default focus on the filter entry but keep it on the plugin list as it was before.

**Comment by eht16** [comment]
Sorry for the unnecessary hbox, this was just a left-over, also the incorrect space value. Fixed in 6b56029.

Do we really want to make the buttons secondary? I found such code in src/search.c:542 but it is a bit ugly as it would require using gtk_dialog_get_action_area() which is already deprecated. This is why I explicitly used the gtk_dialog_add_button() API even it doesn't seem to support something like defining buttons as secondary.
If we really want, I can do, of course.

**Comment by b4n** [comment]
> Do we really want to make the buttons secondary? I found such code in src/search.c:542 but it is a bit ugly as it would require using gtk_dialog_get_action_area() which is already deprecated.

Right, AFAIK this is because of the new buttons in the [header bar thingy](https://developer.gnome.org/gtk3/stable/GtkDialog.html#GtkDialog--use-header-bar).

Well, we don't really have to, it's just nicer (IMHO):

capture du 2015-02-14 15 05 16

but it's true that the code for that would be a bit uglier (though, we could use Glade to build the static part of this dialog too), so I don't know, maybe we don't care.

**Comment by techee** [comment]
Cool! I like the searchbar above the list and all the buttons in the same row too.

I would just suggest the following regarding the search itself: take the input string from filter, split it by whitespace into words and if each of the words appears in the name/description, show it in the list. So for instance for the Commander plugin if you enter "provide command" or any other permutation of the description words into the filter field, it will find the plugin.

Also, I have tried just on osx but after filtering I can't enable/disable the plugin using the checkbox - does this happen on linux too?

**Comment by b4n** [comment]
> Also, I have tried just on osx but after filtering I can't enable/disable the plugin using the checkbox - does this happen on linux too?

Good catch.  The code must be using iters/paths from the filter to search the non-filtered list so they are wrong.

**Comment by techee** [comment]
And one more minor observation - I think the OK button should rather be Close - plugins are loaded immediately after they are selected so the OK button actually doesn't confirm anything.

**Comment by codebrainz** [comment]
It might look a little nicer to remove the`Filter:` label and setting the primary entry icon to a symbolic find icon, and using the placeholder text "Filter plugins list..." or similar. Later on we could replace it with a GtkSearchEntry or whatever.

**Comment by b4n** [comment]
> It might look a little nicer to remove the`Filter:` label and setting the primary entry icon to a symbolic find icon

Yep, I though about that but forgot to mention it.

```c
	gtk_entry_set_icon_from_stock(GTK_ENTRY(filter_entry), GTK_ENTRY_ICON_PRIMARY, GTK_STOCK_FIND);
```

> and using the placeholder text "Filter plugins list..." or similar. Later on we could replace it with a GtkSearchEntry or whatever.

Unfortunately [placeholder text](https://developer.gnome.org/gtk3/stable/GtkEntry.html#gtk-entry-set-placeholder-text) for entries is too recent (3.2) so we cannot depend on it yet.

**Comment by techee** [comment]
I think it would be also nice to search in the plugin's author field - there are some (unnamed) people whose plugins you always want (inspired by the fantastic new Git Change Bar plugin).

And I think the logic in search (apart from what I suggested before) should change - now it searches first in name and if not found, in description - this logic is hidden to users and search results might be confusing for them. I think the search should be performed in the concatenation of name, description and author.

**Comment by elextr** [comment]
Whats the difference between searching in name then description and searching in name concatenated with description?

**Comment by techee** [comment]
Currently it's not searching in name and then description, but

```
if (not found in name)
    search_in_description()
```

This means that for plugins

```
Name: Project Manager
Description: whatever

Name: Project Developer
Description: Project developer is an improved version of project manager
```

it finds just the first plugin but not the second.

**Comment by techee** [comment]
Forgot to mention when searching for "project manager" in the above.

**Comment by techee** [comment]
In addition searching in the concatenated string means that when searching for "project whatever" (and searching for each word separately as I described before), you get the first plugin but if you searched separately for "project whatever" in name and description, you wouldn't get any result.

**Comment by elextr** [comment]
So how will that change if its searching in

    Project Manager whatever
    Project Developer Project developer is an improved version of project manager

It will still find the first before the second.

**Comment by elextr** [comment]
Didn't see your last post until just too late :)

Yeah, searching by separate words is different.  But it shouldn't do that.  If I ask for project manager I want project manager, not project the picture of the manager.  I hate those stupid searches that use separate words by default :(

**Comment by techee** [comment]
Well, this is how google works...

And I think the search in the plugin manager should be targeted at users who don't know exactly the name of the plugin and are searching for some functionality. "git changebar" and "changebar git" should both find the Colomban's new plugin.

**Comment by elextr** [comment]
Sigh dumb it down for the stupidest :(

But it doesn't really matter since we are unlikely to have so infinitely many plugins that search is really *that* useful.  So I will be ok with Enrico doing it whichever way he is happy with.

**Comment by techee** [comment]
With infinitely many plugins the search isn't _that_ useful either - you have a good chance to get infinite number of results :-).

And on the opposite, for the reason there are not so many plugins in Geany I think it's more probable you won't find anything so it's not necessary to be too strict in the search.

**Comment by elextr** [comment]
> With infinitely many plugins the search isn't that useful either - you have a good chance to get infinite number of results :-).

Well, just have infinitely many monkeys coding on infinitely many keyboards ... :-)

> And on the opposite, for the reason there are not so many plugins in Geany I think it's more probable you won't find anything so it's not necessary to be too strict in the search.

Agree, thats why I don't think it matters here.

**Comment by kugel-** [comment]
Am 19.02.2015 um 00:23 schrieb Jiří Techet:
>
> Well, this is how google works...
>

Awesome, you know how Google's algorithm works? :)

I don't know it but even google's search isn't as dumb as plain OR of 
each word. clearly it weights results with multiple matched words 
higher. And it ignores some common words like "and".

Any search that does plain OR without any way to change that (e.g. with 
keywords) is utterly useless because it usually gives too many unrelated 
results.

**Comment by techee** [comment]
> >
> > Well, this is how google works...
> >
>
> Awesome, you know how Google's algorithm works? :)
>
>
Of course just in principle - it finds documents containing _all_ the words
you enter (not talking about sorting, guessing synonyms, adjusting for
language-specific things, etc.)


> I don't know it but even google's search isn't as dumb as plain OR of
> each word. clearly it weights results with multiple matched words
> higher. And it ignores some common words like "and".
>
> Any search that does plain OR without any way to change that (e.g. with
> keywords) is utterly useless because it usually gives too many unrelated
> results.


Totally agree, but I'm talking about AND search all the time (OR is really
useless) - I'm just saying words should be matched separately.

**Comment by eht16** [comment]
Roadmap:
- will fix the "not enough variable arguments to fit a sentinel" warning
- will fix the broken enable/disable of plugins (not that it would be important for that dialog...:D)
- will remove the label and add a primary icon

I will not:
- change the search algorithm for now
- make the bottom buttons secondary

I agree with @techee that first searching the name, then if not found, the description is not the best way. This certainly should be improved. About the other ideas regarding the search algorithm I'm not sure. Basically, it should just be filter to get quick to the desired plugin, not creating another, as it has been mentioned beforehand already, Google :).

We can probably improve the search/filter code after the code has been initially merged and further discussed.

@b4n I'm really unsure about the secondary buttons issue. If you want this change, I can do but since I don't use GTK3 yet, I had no idea about that "header bar thingy" :).

**Comment by techee** [comment]
> Roadmap:
>
>    - will fix the "not enough variable arguments to fit a sentinel"
>    warning
>    - will fix the broken enable/disable of plugins (not that it would be
>    important for that dialog...:D)
>    - will remove the label and add a primary icon
>
> I will not:
>
>    - change the search algorithm for now
>    - make the bottom buttons secondary
>
> I agree with @techee <https://github.com/techee> that first searching the
> name, then if not found, the description is not the best way. This
> certainly should be improved. About the other ideas regarding the search
> algorithm I'm not sure. Basically, it should just be filter to get quick to
> the desired plugin, not creating another, as it has been mentioned
> beforehand already, Google :).
>
> We can probably improve the search/filter code after the code has been
> initially merged and further discussed.
>
> I can provide the "improved searching" patch myself once your patch is
finished and you can decide whether you like it or not. I definitely don't
propose to create new Google, what I suggest is by a rough estimate +10
lines of code and completely trivial.

**Comment by techee** [comment]
OK, this is it:

https://github.com/eht16/geany/pull/1

**Comment by b4n** [comment]
> @b4n I'm really unsure about the secondary buttons issue. If you want this change, I can do […]

Don't bother it's really not very important.

**Comment by eht16** [comment]
I finished my roadmap.

@b4n a) I changed the spacing for the first label and the search entry in the vbox, each to 3. This looks fine to me. Hope this is ok.
b75eb4e should fix the sentinel warning. However, for some reason with all -Wformat warnings enabled I could find, my gcc 4.9.2 didn't issue that warning before.

@techee sorry if my previous comment sounded a bit harsh. I was actually not that concerned about the search algorithm nor that your way isn't appropriate. I just wanted to say to postpone this part until the other issues were resolved.
I merged PR https://github.com/eht16/geany/pull/1 and am happy :). Thanks.

N.B. I will squash the additional commits before merging.

**Comment by techee** [comment]
>
> @techee <https://github.com/techee> sorry if my previous comment sounded
> a bit harsh. I was actually not that concerned about the search algorithm
> nor that your way isn't appropriate. I just wanted to say to postpone this
> part until the other issues were resolved.


Not at all, I was just worried that you think I want to add some search
monster. Glad you like it :-).

**Comment by b4n** [comment]
> @b4n a) I changed the spacing for the first label and the search entry in the vbox, each to 3. This looks fine to me. Hope this is ok.

I wonder why we have spacing=3 on the box and add stuff on the children… what about simply setting spacing=6 and leave the children padding=0?

Doesn't matter much though, as long as the result doesn't look bad ^^

```diff
diff --git a/src/plugins.c b/src/plugins.c
index 844d1a8..4287eed 100644
--- a/src/plugins.c
+++ b/src/plugins.c
@@ -1684,9 +1684,9 @@ static void pm_show_dialog(GtkMenuItem *menuitem, gpointer user_data)
     pm_widgets.popup_help_menu_item = menu_item;
 
     /* put it together */
-    vbox2 = gtk_vbox_new(FALSE, 3);
-    gtk_box_pack_start(GTK_BOX(vbox2), label, FALSE, FALSE, 3);
-    gtk_box_pack_start(GTK_BOX(vbox2), filter_entry, FALSE, FALSE, 3);
+    vbox2 = gtk_vbox_new(FALSE, 6);
+    gtk_box_pack_start(GTK_BOX(vbox2), label, FALSE, FALSE, 0);
+    gtk_box_pack_start(GTK_BOX(vbox2), filter_entry, FALSE, FALSE, 0);
     gtk_box_pack_start(GTK_BOX(vbox2), swin, TRUE, TRUE, 0);
 
     g_signal_connect(pm_widgets.dialog, "response", G_CALLBACK(pm_dialog_response), NULL);
```

> b75eb4e should fix the sentinel warning. However, for some reason with all -Wformat warnings enabled I could find, my gcc 4.9.2 didn't issue that warning before.

That's weird…  I get it 86a530f68ed2a7856488e5d183bca8606f60f4ba even only with `-Wformat` (which is enabled by `-Wall`), and I have 4.9.2 from Debian.

Anyway, yes, it's fixed, thanks :)

**Comment by b4n** [comment]
BTW, it looks/behaves quite neat now :)

(but the buttons are misaligned on GTK3 for some reason… don't worry though, it was like that even before)

**Comment by eht16** [comment]
> I wonder why we have spacing=3 on the box and add stuff on the children… what about simply setting spacing=6 and leave the children padding=0?

I like spacing=6  and padding=0, so committed. Thanks.

Still no idea about the sentinel warning issue. I also have gcc 4.9.2 but from ArchLinux. I just tested again with -Wall, with -Wformat and with many many more warning option enabled, no chance to get this warning. Also searching the net for a *missing* compiler warning is quite hard :).
But well, as long as it is fixed, I'm happy.

Is the button order on GTK 3 just reversed and completely mixed?
Maybe it is related to the response IDs used? The GTK default response IDs are negative while the used IDs for our custom buttons are positive. Maybe this makes a difference for button ordering?

**Comment by b4n** [comment]
> Is the button order on GTK 3 just reversed and completely mixed?

I said misaligned, not "misordered" or something ;)
See for example my capture in https://github.com/geany/geany/pull/414#issuecomment-74376237 to see what I mean: the spacing around the buttons is incorrect with GTK3 (but fine with GTK2).

**Comment by eht16** [comment]
Oops, sorry for not carefully enough reading your comment. No further idea then.

**Comment by eht16** [comment]
I just squashed the smaller commits together, so it is a bit cleaner.
If there are no further problems, I'd like to get it merged.

**Comment by techee** [comment]
@eht16 As I mentioned before (which probably got lost in all the discussions), shouldn't the OK button be renamed to Close? It just closes the dialog and doesn't confirm the dialog or anything like that.

**Comment by techee** [comment]
@eht16 One more thing - when the dialog opens, the Help, Preferences, Keybindings buttons are clickable. Only after you select a row of a non-enabled plugin, they become insensitive. Even though it doesn't seem to cause any trouble, I think they should be insensitive from the beginning.

**Comment by eht16** [comment]
@techee I changed the OK button to Close, thanks for the reminder.

But I don't experience the behaviour you described with the initial state of the Help/Configure/Keybindings buttons. For me, there is always the first plugin selected and the button states represent the selected plugin.
However, maybe it is better to set the initial input focus to the filter entry. At the beginning of the plugin manager dialog improvements, I set the input focus explicitly to the treeview at the time where the patch implemented the interactive treeview search. But since this is gone in the meantime, the input focus might be more useful on the filter entry.
If we agree on this, then the intial button states are definitely wrong and I will correct this.

**Comment by techee** [comment]
@eht16 Interesting. What I see is this:

screenshot - 29 3 2015 - 21 23 08

There is a frame around the checkbox of the first plugin and it apparently has input (if you press space, it gets checked and the row gets highlighted) but the row isn't highlighted initially. Maybe a different GTK version or something like that. So I think there should be an explicit row selection to be sure.

Search having focus sounds like a good idea to me.

**Comment by kugel-** [comment]
Does the search include the filename (xxx.so) somehow? The other day I was searching for a g-p plugin. I knew the filename but not its translated display name.

**Comment by techee** [comment]
Not at the moment but it's trivial to add it (and yeah, sounds like a good idea to me).

**Comment by eht16** [comment]
@techee still no idea what the inital dialog focus behaviour is that different for us :(. I'm using good ol' GTK 2.24.27 with a custom theme (originally based on some Greybird version, IIRC).
Anyway, I changed the input focus to the filter entry and explicitly set an initial button state which is a good thing in any case.
Thanks for noticing.

I personally don't mind the plugin filename search change. If you like to, add it or maybe add it after merging, whatever you prefer.

**Comment by eht16** [comment]
https://github.com/eht16/geany/pull/2 merged and commits squashed.

**Comment by eht16** [comment]
Anything left to do? Otherwise I would like to merge this PR.

**Comment by b4n** [comment]
I didn't follow all the discussion closely, but LGTM and it's a great improvement :)

**Comment by eht16** [comment]
Good catch, break added.

**Comment by b4n** [comment]
LGTM :)
